content/en/blog/_posts/2019-12-09-volume-snapshot-beta.md

```diff
@@ -0,0 +1,453 @@
+
+As of the publishing of this blog, the following CSI drivers have been updated to support volume snapshots beta:
+
+- [GCE Persistent Disk CSI Driver](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver)
+- [Portworx CSI Driver](https://github.com/libopenstorage/openstorage/tree/master/csi)
+- [NetApp Trident CSI Driver](https://github.com/NetApp/trident)
+
+Beta level Volume Snapshot support for other [CSI drivers](https://kubernetes-csi.github.io/docs/drivers.html) is pending, and should be available soon.
+
+## Kubernetes Volume Snapshot Beta API
+
+A number of changes were made to the Kubernetes volume snapshot API between alpha to beta. These changes are not backward compatible. The purpose of these changes was to make API definitions clear and easier to use.
+
+The following changes are made:
+
+- `DeletionPolicy` is now a required field rather than optional in both `VolumeSnapshotClass` and `VolumeSnapshotContent`. This way the user has to explicitly specify it, leaving no room for confusion.
+- `VolumeSnapshotSpec` has a new required `Source` field. `Source` may be either a `PersistentVolumeClaimName` (if dynamically provisioning a snapshot) or `VolumeSnapshotContentName` (if pre-provisioning a snapshot).
+- `VolumeSnapshotContentSpec` also has a new required `Source` field. This `Source` may be either a `VolumeHandle` (if dynamically provisioning a snapshot) or a `SnapshotHandle` (if pre-provisioning volume snapshots).
+- `VolumeSnapshotStatus` now contains a `BoundVolumeSnapshotContentName` to indicate the `VolumeSnapshot` object is bound to a `VolumeSnapshotContent`.
+- `VolumeSnapshotContent`now contains a `Status` to indicate the current state of the content. It has a field `SnapshotHandle` to indicate that the `VolumeSnapshotContent` represents a snapshot on the storage system.
+
+The beta Kubernetes VolumeSnapshot API object:
+
+```go
+type VolumeSnapshot struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+
+        Spec VolumeSnapshotSpec
+        Status *VolumeSnapshotStatus
+}
+```
+
+```go
+type VolumeSnapshotSpec struct {
+	Source VolumeSnapshotSource
+	VolumeSnapshotClassName *string
+}
+// Exactly one of its members MUST be specified
+type VolumeSnapshotSource struct {
+	// +optional
+	PersistentVolumeClaimName *string
+	// +optional
+	VolumeSnapshotContentName *string
+}
+```
+
+```go
+type VolumeSnapshotStatus struct {
+	BoundVolumeSnapshotContentName *string
+	CreationTime *metav1.Time
+	ReadyToUse *bool
+	RestoreSize *resource.Quantity
+	Error *VolumeSnapshotError
+}
+```
+
+
+The beta Kubernetes VolumeSnapshotContent API object:
+
+```go
+type VolumeSnapshotContent struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+
+        Spec VolumeSnapshotContentSpec
+        Status *VolumeSnapshotContentStatus
+}
+```
+
+```go
+type VolumeSnapshotContentSpec struct {
+         VolumeSnapshotRef core_v1.ObjectReference
+         PersistentVolumeRef core_v1.ObjectReference
```

Comment:
This one was also crossed out in the original google doc.  It should just be removed.